spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java

```diff
@@ -194,21 +172,64 @@ public RabbitConsumerProperties getExtendedConsumerProperties(String channelName
 		return this.extendedBindingProperties.getExtendedConsumerProperties(channelName);
 	}
 
-
 	@Override
 	public RabbitProducerProperties getExtendedProducerProperties(String channelName) {
 		return this.extendedBindingProperties.getExtendedProducerProperties(channelName);
 	}
 
 	@Override
-	protected MessageProducer createConsumerEndpoint(String name, String group, Queue destination,
-			ExtendedConsumerProperties<RabbitConsumerProperties> properties) {
+	protected MessageHandler createProducerMessageHandler(final ProducerDestination producerDestination,
+														ExtendedProducerProperties<RabbitProducerProperties> producerProperties)
+			throws Exception {
+		String prefix = producerProperties.getExtension().getPrefix();
+		String exchangeName = producerDestination.getName();
+		String destination = StringUtils.isEmpty(prefix) ? exchangeName : exchangeName.substring(prefix.length());
+		final AmqpOutboundEndpoint endpoint = new AmqpOutboundEndpoint(buildRabbitTemplate(producerProperties.getExtension()));
+		endpoint.setExchangeName(producerDestination.getName());
+		RabbitProducerProperties extendedProperties = producerProperties.getExtension();
+		String routingKeyExpression = extendedProperties.getRoutingKeyExpression();
+		if (!producerProperties.isPartitioned()) {
+			if (routingKeyExpression == null) {
+				endpoint.setRoutingKey(destination);
+			}
+			else {
+				endpoint.setRoutingKeyExpressionString(routingKeyExpression);
+			}
+		}
+		else {
+			if (routingKeyExpression == null) {
+				endpoint.setRoutingKeyExpressionString(buildPartitionRoutingExpression(destination));
+			}
+			else {
+				endpoint.setRoutingKeyExpressionString(buildPartitionRoutingExpression(routingKeyExpression));
```

Comment:
OK.
Looks I configure `routingKeyExpression = "payload.toUpperCase()"`.
When we come to the current logic and let's say `1` for partition we end up with the : 
```
 payload.toUpperCase()-1
```
just because this: ` "'" + expressionRoot + "-'`.
No we don't evaluate that `routingKeyExpression` here.